File: libs/nox-ecs/examples/six_dof.rs
use elodin_conduit::well_known::{Material, Mesh};
use nox::{nalgebra, SpatialForce, SpatialInertia, SpatialTransform};
use nox::{nalgebra::vector, SpatialMotion};
use nox_ecs::{six_dof::*, spawn_tcp_server, Query, WorldPos};
use nox_ecs::{World, WorldBuilder};

fn gravity(pos: Query<(WorldPos, Inertia, Force)>) -> Query<Force> {
    const G: f64 = 6.649e-11;
    let big_m: f64 = 1.0 / G;
    pos.map(|world_pos: WorldPos, inertia: Inertia, force: Force| {
        let mass = inertia.0.mass();
        let r = world_pos.0.linear();
        let norm = r.clone().norm();
        let force = force.0
            + SpatialForce::from_linear(
                -r / (norm.clone() * norm.clone() * norm) * G * big_m * mass,
            );
        Force(force)
    })
    .unwrap()
}

fn main() {
    tracing_subscriber::fmt::init();
    let mut world = World::default();
    let model = world.insert_asset(Mesh::sphere(0.1, 36, 18));
    let material = world.insert_asset(Material::color(1.0, 1.0, 1.0));

    world.spawn(Body {
        pos: WorldPos(SpatialTransform {
            inner: vector![1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0].into(),
        }),
        vel: WorldVel(SpatialMotion {
            inner: vector![0.0, 0.0, 0.0, 0.0, 0.0, 1.0].into(),
        }),
        accel: WorldAccel(SpatialMotion {
            inner: vector![0.0, 0.0, 0.0, 0.0, 0.0, 0.0].into(),
        }),
        model,
        material,
        force: Force(SpatialForce {
            inner: vector![0.0, 0.0, 0.0, 0.0, 0.0, 0.0].into(),
        }),
        mass: Inertia(SpatialInertia {
            inner: vector![1.0, 1.0, 1.0, 0.0, 0.0, 0.0, 1.0].into(),
        }),
    });
    let builder = WorldBuilder::new(world, six_dof(|| gravity, 1.0 / 60.0));
    let client = nox::Client::cpu().unwrap();
    let exec = builder.build(&client).unwrap();
    spawn_tcp_server(
        "0.0.0.0:3104".parse().unwrap(),
        exec,
        &client,
        std::time::Duration::from_secs_f64(1.0 / 60.0),
    )
    .unwrap();
}

File: libs/nox-py/examples/six_dof.py
from jax import numpy as np
from jax.numpy import linalg as la
from elodin import *

def gravity_impl(pos, inertia, force):
  G = 6.67430e-11
  big_m = 1/G
  mass = inertia.mass()
  r = pos.linear()
  norm = la.norm(r)
  f = -r / norm * norm * norm * G * big_m * mass
  return Force.from_linear(f)

@system
def gravity(q: Query[WorldPos, Inertia, Force]) -> ComponentArray[Force]:
  return q.map(Force, gravity_impl)

w = WorldBuilder()
w.spawn(Body(
    world_pos = WorldPos.from_linear(np.array([0.,0.,0.01])),
    world_vel = WorldVel.from_linear(np.array([0.,0.,0.])),
    inertia = Inertia.from_mass(1.0),
    mesh = w.insert_asset(Mesh.sphere(1)),
    material = w.insert_asset(Material.color(25.3, 18.4, 1.0))
))
w.spawn(Body(
    world_pos = WorldPos.from_linear(np.array([5.,0.,0.])),
    world_vel = WorldVel.from_linear(np.array([0.,0.,10.])),
    inertia = Inertia.from_mass(1.0),
    mesh = w.insert_asset(Mesh.sphere(0.2)),
    material = w.insert_asset(Material.color(1.0, 1.0, 1.0))
))
w.spawn(Body(
    world_pos = WorldPos.from_linear(np.array([8.,0.,0.])),
    world_vel = WorldVel.from_linear(np.array([0.,0.,24.])),
    inertia = Inertia.from_mass(2.0),
    mesh = w.insert_asset(Mesh.sphere(0.3)),
    material = w.insert_asset(Material.color(1.0, 1.0, 1.0))
))
sys = six_dof(1.0 / 60.0, gravity)
sim = w.run(sys)
